crates/util/src/fibonacci_backoff.rs

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::time::Duration;
+
+use backoff::backoff::Backoff;
+
+// Fibonacci-based backoff delay intervals capped at 5 mins
+const BACKOFF_INTERVALS_MS: [u64; 14] = [
+    1000, 1000, 2000, 3000, 5000, 8000, 13000, 21000, 34000, 55000, 89000, 144_000, 233_000,
+    300_000,
+];
+
+#[derive(Debug)]
+pub struct FibonacciBackoff {
+    num_retries: usize,
+    pub randomization_factor: f64,
+    pub max_retries: Option<usize>,
+}
+
+impl Default for FibonacciBackoff {
+    fn default() -> FibonacciBackoff {
+        FibonacciBackoff {
+            num_retries: 0,
+            randomization_factor: 0.3,
+            max_retries: None,
+        }
+    }
+}
+
+impl Backoff for FibonacciBackoff {
+    fn reset(&mut self) {
+        self.num_retries = 0;
+    }
+
+    fn next_backoff(&mut self) -> Option<Duration> {
+        self.num_retries += 1;
+
+        if let Some(max_retries) = self.max_retries {
+            if self.num_retries > max_retries {
+                return None;
+            }
+        }
+
+        let interval = if self.num_retries >= BACKOFF_INTERVALS_MS.len() {
+            Duration::from_millis(BACKOFF_INTERVALS_MS[BACKOFF_INTERVALS_MS.len() - 1])
+        } else {
+            Duration::from_millis(BACKOFF_INTERVALS_MS[self.num_retries])
+        };
+
+        let randomized_interval = get_random_value_from_interval(
+            self.randomization_factor,
+            rand::random::<f64>(),
+            interval,
+        );
+
+        Some(randomized_interval)
+    }
+}
+
+fn get_random_value_from_interval(
+    randomization_factor: f64,
+    random: f64,
+    current_interval: Duration,
+) -> Duration {
+    let current_interval_nanos = duration_to_nanos(current_interval);
+
+    let delta = randomization_factor * current_interval_nanos;
```

Comment:
This uses Backoff library code that is used for exponential backoff but is not exposed publicly
https://github.com/ihrwein/backoff/blob/587e2da8fb2dcfc65ca544cb9249022c51f1406e/src/exponential.rs#L61